components/style/properties/gecko.mako.rs

```diff
@@ -1671,6 +1671,54 @@ fn static_assert() {
 
     <%call expr="impl_coord_copy('word_spacing', 'mWordSpacing')"></%call>
 
+    fn clear_text_emphasis_style_if_string(&mut self) {
+        use nsstring::nsString;
+        if self.gecko.mTextEmphasisStyle == structs::NS_STYLE_TEXT_EMPHASIS_STYLE_STRING as u8 {
+            self.gecko.mTextEmphasisStyleString.assign(&nsString::new());
+            self.gecko.mTextEmphasisStyle = structs::NS_STYLE_TEXT_EMPHASIS_STYLE_NONE as u8;
+        }
+    }
+
+    pub fn set_text_emphasis_style(&mut self, v: longhands::text_emphasis_style::computed_value::T) {
+        use nsstring::nsCString;
+        use properties::longhands::text_emphasis_style::computed_value::T;
+        use properties::longhands::text_emphasis_style::ShapeKeyword;
+
+        self.clear_text_emphasis_style_if_string();
+        let te = match v {
+            T::None => structs::NS_STYLE_TEXT_EMPHASIS_STYLE_NONE,
+            T::Keyword(keyword) => {
+                let fill = if keyword.fill {
+                    structs::NS_STYLE_TEXT_EMPHASIS_STYLE_FILLED
+                } else {
+                    structs::NS_STYLE_TEXT_EMPHASIS_STYLE_OPEN
+                };
+                let shape = match keyword.shape {
+                    ShapeKeyword::Dot => structs::NS_STYLE_TEXT_EMPHASIS_STYLE_DOT,
+                    ShapeKeyword::Circle => structs::NS_STYLE_TEXT_EMPHASIS_STYLE_CIRCLE,
+                    ShapeKeyword::DoubleCircle => structs::NS_STYLE_TEXT_EMPHASIS_STYLE_DOUBLE_CIRCLE,
+                    ShapeKeyword::Triangle => structs::NS_STYLE_TEXT_EMPHASIS_STYLE_TRIANGLE,
+                    ShapeKeyword::Sesame => structs::NS_STYLE_TEXT_EMPHASIS_STYLE_SESAME,
+                };
```

Comment:
Like @upsuper said, we need to always set the string here. The list of constants is at https://dxr.mozilla.org/mozilla-central/source/layout/style/nsRuleNode.cpp#4394. It's probably best to add a `.char()` method to `shape` that returns these. Later I'll fix the bindings generation so that we can use the `kTextEmphasisChars` table directly.